### PR TITLE
Support UTIME_OMIT/UTIME_NOW in futimens

### PR DESCRIPTION
**Pull request by sbc100**
Fixes: #22348

### CODE REVIEW

**Comment by sbc100** [comment]
Updated to include wasmfs support

File: src/library_syscall.js [hunk 1]
```diff
@@ -56,11 +56,11 @@ var SyscallsLibrary = {
...
       var ctime = stat.ctime.getTime();
       {{{ makeSetValue('buf', C_STRUCTS.stat.st_atim.tv_sec, 'Math.floor(atime / 1000)', 'i64') }}};
-      {{{ makeSetValue('buf', C_STRUCTS.stat.st_atim.tv_nsec, '(atime % 1000) * 1000', SIZE_TYPE) }}};
+      {{{ makeSetValue('buf', C_STRUCTS.stat.st_atim.tv_nsec, '(atime % 1000) * 1000 * 1000', SIZE_TYPE) }}};
       {{{ makeSetValue('buf', C_STRUCTS.stat.st_mtim.tv_sec, 'Math.floor(mtime / 1000)', 'i64') }}};
-      {{{ makeSetValue('buf', C_STRUCTS.stat.st_mtim.tv_nsec, '(mtime % 1000) * 1000', SIZE_TYPE) }}};
+      {{{ makeSetValue('buf', C_STRUCTS.stat.st_mtim.tv_nsec, '(mtime % 1000) * 1000 * 1000', SIZE_TYPE) }}};
       {{{ makeSetValue('buf', C_STRUCTS.stat.st_ctim.tv_sec, 'Math.floor(ctime / 1000)', 'i64') }}};
-      {{{ makeSetValue('buf', C_STRUCTS.stat.st_ctim.tv_nsec, '(ctime % 1000) * 1000', SIZE_TYPE) }}};
+      {{{ makeSetValue('buf', C_STRUCTS.stat.st_ctim.tv_nsec, '(ctime % 1000) * 1000 * 1000', SIZE_TYPE) }}};
```
Comment (kripken) [inline src/library_syscall.js:63]: Are these separate fixes, or related? (they do look correct)

File: src/library_syscall.js [hunk 2]
```diff
@@ -978,19 +978,34 @@ var SyscallsLibrary = {
...
+      atime = now;
+      mtime = now;
     } else {
       var seconds = {{{ makeGetValue('times', C_STRUCTS.timespec.tv_sec, 'i53') }}};
       var nanoseconds = {{{ makeGetValue('times', C_STRUCTS.timespec.tv_nsec, 'i32') }}};
-      atime = (seconds*1000) + (nanoseconds/(1000*1000));
+      if (nanoseconds == {{{ cDefs.UTIME_NOW }}}) {
+        atime = now;
+      } else if (nanoseconds == {{{ cDefs.UTIME_OMIT }}}) {
+        atime = -1;
```
Comment (kripken) [inline src/library_syscall.js:991]: Is `-1`'s role obvious here? I gather from the code that `FS.utime` just takes the max of the times, so `-1` being negative is enough to make the other value be chosen, but I had to go read `FS.utime` for that - maybe a comment here would be good?

File: system/lib/wasmfs/syscalls.cpp [hunk 3]
```diff
@@ -1123,6 +1119,16 @@ int __syscall_readlinkat(int dirfd,
   return bytes;
 }
 
+static double timespec_to_ms(timespec ts) {
+  if (ts.tv_nsec == UTIME_OMIT) {
+    return INFINITY;
```
Comment (kripken) [inline system/lib/wasmfs/syscalls.cpp:1124]: Where is INFINITY defined?

File: test/utime/test_futimens.c [hunk 4]
```diff
@@ -31,6 +32,23 @@ void setup() {
   symlink("file", "folder/file-link");
 }
 
+void check_times(int fd, struct timespec* expected, int tollerance) {
```
Comment (kripken) [inline test/utime/test_futimens.c:35]:
```suggestion
void check_times(int fd, struct timespec* expected, int tolerance) {
```

File: test/utime/test_futimens.c [hunk 5]
```diff
@@ -54,13 +72,67 @@ void test() {
...
+  printf("check single UTIME_OMIT...\n");
+  newtimes[0].tv_sec = 41;
+  newtimes[0].tv_nsec = UTIME_OMIT;
+  newtimes[1].tv_sec = 42;
+  newtimes[1].tv_nsec = 88;
+  err = futimens(fd, newtimes);
+  assert(!err);
+
+#if defined(__EMSCRIPTEN__) && !defined(WASMFS)
+  // The trandiation emscripten FS only supports single timestamp so both
```
Comment (kripken) [inline test/utime/test_futimens.c:109]:
```suggestion
  // The original emscripten FS (in JS) only supports a single timestamp so both
```

File: src/library_syscall.js (hunk 1 above)
Comment (sbc100) [reply src/library_syscall.js:63]: I noticed these values were off by 1000 while working on the test case.  I guess its a separate fix but I think its fine to include it here.

File: system/lib/wasmfs/syscalls.cpp (hunk 3 above)
Comment (sbc100) [reply system/lib/wasmfs/syscalls.cpp:1124]: [<math.h>: ](https://en.cppreference.com/w/c/numeric/math/INFINITY)

Comment (kripken) [reply system/lib/wasmfs/syscalls.cpp:1124]: Ah, thanks, what I was confused about is that `tv_nsec` is a `long`, and I missed that we return a different type here, `double`.

**Review by kripken — APPROVED** [review]

**Pull request merged**